### PR TITLE
[Resources] [ARM] Add cmdlets for ARM template What-If.

### PR DESCRIPTION
**Pull request by shenglol**
## Description

- Added two new cmdles for ARM template What-If operation:
    - `Get-AzDeploymentWhatIf` (alias: `Get-AzSubscriptionDeploymentWhatIf`)
    - `Get-AzResourceGroupDeploymentWhatIf`

- Overrode `-WhatIf` and `-Confirm` switch parameters for `New-AzDeploymentWhatIf` and `New-AzResourceGroupWhatIf` to use ARM template What-If output.

**Notes for code reviewers**:
We would like to get a preview release for `Az.Resources` at the end of this month. There are some improvements we need to make on our back-end before we can merge this PR.

## Checklist

- [x] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [x] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [x] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

**Comment by adxsdkps** [comment]
Can one of the admins verify this patch?

**Comment by kilasuit** [comment]
Do not merge this

You should not ever suffix a command with WhatIf and the design of this, and its functionality realistically should be a component of the core Test commands `Test-AzDeployment` `Test-AzManagementGroupDeployment` `Test-AzResourceGroupDeployment` or `Test-AzTenantDeployment` not a Get Command

**Comment by shenglol** [comment]
> Please Do not merge this
> 
> You should not ever suffix a command with WhatIf and the design of this, and its functionality realistically should be a component of the core Test commands `Test-AzDeployment` `Test-AzManagementGroupDeployment` `Test-AzResourceGroupDeployment` or `Test-AzTenantDeployment` not a Get Command
> 
> I would strongly urge you redesign this as functionality into the above mentioned commands cc @markcowl @sphibbs

Hi @kilasuit , thanks for the feedback.

For your reference, [ARM Template deployment What-If](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-deploy-what-if) is a feature currently in private preview. It basically tells you what would happen when you deploy a template if it is valid. The `Test-*AzDeployment` cmdlets serve a different purpose. It will only validate a template and its parameters.

We already had a internal discussion with @markcowl and decided to go with a separate cmdlet. There's an existing issue https://github.com/Azure/azure-powershell/issues/10619 about the cmdlet name, and the `Get` verb is actually based on suggestions we got there.

cc @alex-frankel @majastrz .

**Comment by alex-frankel** [comment]
I don't have a strong opinion on Get vs Test, but this is not what Test returns today. Should we do Test-AzDeployment -WhatIf to return the object? Will that be confusing that -WhatIf for `Test-Az*Deployment` returns an object while -WhatIf for `New-Az*Deployment` returns text?

**Comment by majastrz** [comment]
It sounds like the problem is with `WhatIf` not being a noun. @kilasuit, how about calling that cmdlet `Get-Az*DeploymentWhatIfResult`.

**Comment by shenglol** [comment]
Hi @isra-fel , do you mind taking a look since you reviewed it before? There's not many changes compared to the version that got reverted a few months before.

**Comment by kilasuit** [comment]
> It sounds like the problem is with WhatIf not being a noun. @kilasuit, how about calling that cmdlet Get-Az*DeploymentWhatIfResult.

I personally don't like that all the much more than what is proposed already, but I do prefer that to the current proposal as it reads better,
 
> I don't have a strong opinion on Get vs Test, but this is not what Test returns today. Should we do Test-AzDeployment -WhatIf to return the object? Will that be confusing that -WhatIf for `Test-Az*Deployment` returns an object while -WhatIf for `New-Az*Deployment` returns text?

See this makes makes more sense when you think what it is the end user is doing, in this case they would want to test their current templates will deploy and then optionally test what the template they are going to deploy are going to change from what is already deployed. Which to me is an extension to the Test cmdlets and not a new cmdlet completelty. Therefore I think that overloading the `WhatIf` on `Test` to return an object, at least to me feels more natural & optimal for those running this process in the console rather than having to resort to cmdlet change.

The optimal flow (simplified), in my opinion when working from the console would look like this

```powershell
Test-AzDeployment
Test-AzDeployment -WhatIf
New-AzDeployment
``
Which is only a few actual key presses (thanks to console history and tab completion)

 compared to this 
```powershell
Test-AzDeployment
Get-AzDeploymentWhatIfResult.
New-AzDeployment
```
Hope this helps explain my view on this

**Comment by bergmeister** [comment]
The -WhatIf switch should be on the New-AzDeployment cmdlet, not its Test-AzDeployment counterpart. As a user the -WhatIf switch tells me what would happen if I executed the cmdlets without the -WhatIf switch, hence why it's more appropriate on the New- cmdlet rather than the Test- one.

**Comment by kilasuit** [comment]
> The -WhatIf switch should be on the New-AzDeployment cmdlet, not its Test-AzDeployment counterpart. As a user the -WhatIf switch tells me what would happen if I executed the cmdlets without the -WhatIf switch, hence why it's more appropriate on the New- cmdlet rather than the Test- one.

Generally in other areas I would agree with you, however as we have `Test` cmdlets in this module I think this feels more appropriate on them than on `New` because you are not technically initiating a full new deployment but are testing what would change to already existing deployed resources if you did, hence why I think its should be on Test. 

Though I suppose that you could argue that this functionality it should actually be on both `New` & `Test` cmdlets, though I know that generally I would be using `WhatIf` in template testing stages (i.e. local development) not template deployment via my CD Pipelines. 
The difference in mindsets in the two, whilst subtle, make it worth while having this functionality in both for authors to pick which they feel is more appropriate.

But it definitely should not be another _unnecessary_ cmdlet

**Comment by shenglol** [comment]
> > The -WhatIf switch should be on the New-AzDeployment cmdlet, not its Test-AzDeployment counterpart. As a user the -WhatIf switch tells me what would happen if I executed the cmdlets without the -WhatIf switch, hence why it's more appropriate on the New- cmdlet rather than the Test- one.
> 
> Generally in other areas I would agree with you, however as we have `Test` cmdlets in this module I think this feels more appropriate on them than on `New` because you are not technically initiating a full new deployment but are testing what would change to already existing deployed resources if you did, hence why I think its should be on Test.
> 
> Though I suppose that you could argue that this functionality it should actually be on both `New` & `Test` cmdlets, though I know that generally I would be using `WhatIf` in template testing stages (i.e. local development) not template deployment via my CD Pipelines.
> The difference in mindsets in the two, whilst subtle, make it worth while having this functionality in both for authors to pick which they feel is more appropriate.
> 
> But it definitely should not be another _unnecessary_ cmdlet

The separate cmdlet is useful for users who what to parse the what-if result programmatically, e.g.:
```pwsh
$whatIfResult = Get-AzResourceGroupWhatIf ...
# do whatever you want with $whatIfResult
```

**Comment by brwilkinson** [comment]
I would just like one single cmdlet for everything.

New-AzDeployment

Then depending on the parameters it will deploy at different scopes.... then depending on -whatif or not it will actually deploy or tell me the results of the whatif.

I prefer no separate cmdlets for different deployment scope deployments and no different cmdlets for a test deploy.

It's MUCH easier to modify a set of parameters that can be splatted into the single Cmdlet than having to call the different Cmdlets for the different scenarios.

**Comment by alex-frankel** [comment]
@kilasuit and @bergmeister: 

The clean, text-based WhatIf response (which includes the validation check that the Test* does) will trigger when you do `New-Az*Deployment -WhatIf` or `New-Az*Deployment -Confirm`. So we are expecting most dev/test loops to simply be:

`New-Az*Deployment -Confirm`

or

`New-Az*Deployment -WhatIf`
`New-Az*Deployment`

Personally, I have been enjoying just using `-Confirm` and, depending on the response, I select `Y` or `N`

Only if you need to *programmatically* interrogate the results, you would do `Get-Az*DeploymentWhatIfResult`


@brwilkinson - we did start by trying to unify this all into New-AzDeployment with different parameter sets, but it made the cmdlet autocomplete experience somewhat confusing in certain edge cases because the parameter sets are too different (e.g. Sub, Mg and Tenant deployments all require a location, but RG does not)

We're not against unifying, but we need to spend more time getting the usability right.

**Comment by alex-frankel** [comment]
@isra-fel - can you or someone on your team please take a look at this?

File: src/Resources/Resources/ChangeLog.md [hunk 1]
```diff
@@ -30,6 +30,10 @@
     - Update-AzTag -ResourceId
 * Brought ScopedDeployment from SDK 3.3.0 
 * Breaking changes for policy cmdlets to support typed pipeline objects
+* Add Get-AzDeploymentWhatIf cmdlet for getting ARM template What-If at subscription scope.
```
Comment (isra-fel) [inline src/Resources/Resources/ChangeLog.md]: Please move new contents after "Upcoming Release", our tools for release depend on this structure

Comment (shenglol) [reply src/Resources/Resources/ChangeLog.md]: Missed this one after resolving merge conflicts. Pushed out a fix.

**Comment by isra-fel** [comment]
Hi, we have a commit https://github.com/Azure/azure-powershell/commit/bc5776362805015341494addfa9432633a2442c0 on master to update the mac os image version in our CI checks. The resources preview branch hasn't sync with master for quite some time, that's why the CI is constantly failing.
The simplest way I think is to cherry-pick that commit to your branch. Could you do that? Thanks

**Comment by shenglol** [comment]
> Hi, we have a commit [bc57763](https://github.com/Azure/azure-powershell/commit/bc5776362805015341494addfa9432633a2442c0) on master to update the mac os image version in our CI checks. The resources preview branch hasn't sync with master for quite some time, that's why the CI is constantly failing.
> The simplest way I think is to cherry-pick that commit to your branch. Could you do that? Thanks

That's good to know! I've cherry-picked the commit.

**Comment by shenglol** [comment]
Hi @isra-fel, can you please merge this?

**Pull request merged**